### PR TITLE
Add DistSQL LOCK CLUSTER

### PR DESCRIPTION
**Pull request by Pace2Car**
For #24338.

Changes proposed in this pull request:
  - Add DistSQL LOCK CLUSTER，support stop write, stop read-write lock 
  - Add related test case
  - Update document

---

Before committing this PR, I'm sure that I have checked the following options:
- [x] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [x] I have self-reviewed the commit code.
- [ ] I have (or in comment I request) added corresponding labels for the pull request.
- [x] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [x] I have added corresponding unit tests for my changes.

### CODE REVIEW

File: infra/common/src/main/java/org/apache/shardingsphere/infra/state/cluster/ClusterStateContext.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    
+    @Getter
+    private ClusterStateType currentState = ClusterStateType.OK;
+    
+    /**
+     * Switch state.
+     * 
+     * @param status status
+     */
+    public void switchState(final ClusterStateType status) {
```
Comment (RaigorJiang) [inline infra/common/src/main/java/org/apache/shardingsphere/infra/state/cluster/ClusterStateContext.java:35]: It seems that `state` and `status` express the same meaning, can they be named uniformly?

File: infra/common/src/main/java/org/apache/shardingsphere/infra/state/cluster/ClusterStateContext.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        if (currentState == status) {
+            return;
+        }
+        if (currentState != ClusterStateType.OK && ClusterStateType.OK != status) {
```
Comment (RaigorJiang) [inline infra/common/src/main/java/org/apache/shardingsphere/infra/state/cluster/ClusterStateContext.java:39]: Enum should be on the left

File: infra/common/src/main/java/org/apache/shardingsphere/infra/state/cluster/ClusterStateType.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.shardingsphere.infra.state.cluster;
+
+/**
+ * Cluster state type.
+ */
+public enum ClusterStateType {
```
Comment (RaigorJiang) [inline infra/common/src/main/java/org/apache/shardingsphere/infra/state/cluster/ClusterStateType.java:23]: `ClusterState` may be more accurate than `ClusterStateType`

File: proxy/backend/spi/pom.xml [hunk 4]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (RaigorJiang) [inline proxy/backend/spi/pom.xml:1]: Is it necessary to add this module `spi`

File: infra/common/src/main/java/org/apache/shardingsphere/infra/state/cluster/ClusterStateContext.java (hunk 1 above)
Comment (Pace2Car) [reply infra/common/src/main/java/org/apache/shardingsphere/infra/state/cluster/ClusterStateContext.java:35]: I have fixed this, please review again.

File: proxy/backend/spi/pom.xml (hunk 4 above)
Comment (Pace2Car) [reply proxy/backend/spi/pom.xml:1]: Yes, it is used to support various types of locking and checking operations.

File: proxy/backend/core/src/main/java/org/apache/shardingsphere/proxy/backend/state/impl/ReadOnlyProxyState.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    
+    @Override
+    public void check(final SQLStatement sqlStatement) {
+        if (isWriteStatement(sqlStatement)) {
+            throw new ReadOnlyException();
+        }
+    }
+    
+    private boolean isWriteStatement(final SQLStatement sqlStatement) {
+        return !(sqlStatement instanceof SelectStatement) && !(sqlStatement instanceof ShowStatement) && !(sqlStatement instanceof MySQLShowDatabasesStatement);
```
Comment (RaigorJiang) [inline proxy/backend/core/src/main/java/org/apache/shardingsphere/proxy/backend/state/impl/ReadOnlyProxyState.java:40]: @Pace2Car What about DistSQLStatement?

File: proxy/backend/core/src/main/java/org/apache/shardingsphere/proxy/backend/state/impl/UnavailableProxyState.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    
+    @Override
+    public void check(final SQLStatement sqlStatement) {
+        if (isUnsupportedStatement(sqlStatement)) {
+            throw new UnavailableException();
+        }
+    }
+    
+    private boolean isUnsupportedStatement(final SQLStatement sqlStatement) {
+        return sqlStatement instanceof DMLStatement || sqlStatement instanceof DDLStatement || sqlStatement instanceof RALStatement;
```
Comment (RaigorJiang) [inline proxy/backend/core/src/main/java/org/apache/shardingsphere/proxy/backend/state/impl/UnavailableProxyState.java:40]: Why check RAL but not RDL here  🤔

File: proxy/backend/core/src/main/java/org/apache/shardingsphere/proxy/backend/state/impl/ReadOnlyProxyState.java (hunk 5 above)
Comment (Pace2Car) [reply proxy/backend/core/src/main/java/org/apache/shardingsphere/proxy/backend/state/impl/ReadOnlyProxyState.java]: Thanks for point out, I have fixed.

**Review by RaigorJiang — APPROVED** [review]

**Pull request merged**